modules/dnn/src/onnx/onnx_importer.cpp

```diff
@@ -840,27 +848,68 @@ void ONNXImporter::populateNet(Net dstNet)
             }
         }
 
+
         int id = dstNet.addLayer(layerParams.name, layerParams.type, layerParams);
         layer_id.insert(std::make_pair(layerParams.name, LayerInfo(id, 0)));
 
+        // If no unsupported layers have been found
+        if (op_error.empty())
+        {
+            std::vector<MatShape> layerInpShapes, layerOutShapes, layerInternalShapes;
+            for (int j = 0; j < node_proto.input_size(); j++) {
+                layerId = layer_id.find(node_proto.input(j));
+                if (layerId != layer_id.end()) {
+                    dstNet.connect(layerId->second.layerId, layerId->second.outputId, id, j);
+                    // Collect input shapes.
+                    shapeIt = outShapes.find(node_proto.input(j));
+                    CV_Assert(shapeIt != outShapes.end());
+                    layerInpShapes.push_back(shapeIt->second);
+                }
+            }
+            try
+            {
+                // Compute shape of output blob for this layer.
+                Ptr<Layer> layer = dstNet.getLayer(id);
+                layer->getMemoryShapes(layerInpShapes, 0, layerOutShapes, layerInternalShapes);
+                CV_Assert(!layerOutShapes.empty());
+                outShapes[layerParams.name] = layerOutShapes[0];
+            }
+            catch(cv::Exception& e)
+            {
+                op_error.push_back(e.what());
```

Comment:
I'm completely against that.
You loosing any useful information from C++ exception in that construction.
I didn't see any SDK/library whom uses exceptions in that way.

Reply:
@alalek What design do you want without totally broke the importer?
 We want to give to the user full visibility on all the missing stuffs when he try to import an onnx model and not to let the user to die fixing a single op after  op after unsupport params etc.. We want to give him full visibility of the work that it is required to run that model.

Reply:
There are different levels where assertion appears. Most of them in layers implementation. As I initially proposed in https://github.com/opencv/opencv/issues/14772#issuecomment-500448114 we need to collect just names of unsupported layers. It might be done in in `setUp` method in dnn.cpp:

```
clear();

// Iterate over layers here and summarize the types of unsupported layers

allocateLayers(blobsToKeep_);
```

We couldn't keep all the error messages in a single message. But at least can report about unknown layers types once.

Reply:
I thought you proposed changing the [getLayerInstance](https://github.com/opencv/opencv/blob/6389303c1c3b7269aa805cc35f691dd20fc6dde0/modules/dnn/src/dnn.cpp#L525) function, which is the one that throws the error.

Reply:
@dvd42, Something like
```cpp
std::string errMsg = "";
for (MapIdToLayerData::iterator it = layers.begin(); it != layers.end(); ++it)
{
    LayerData &ld = it->second;
    try
    {
       ld.getLayerInstance()
    }
    catch(...)
    {
        errMsg += ld.type + " (" + ld.name + ");
    }
}
if (!errMsg.empty())
    CV_Error(Error::StsNotImplemented, "Unsupported layers types: " + errMsg);
```

`allocateLayers` is the first method which calls `getLayerInstance` for layers. So we can call it before to collect unsupported types and names.

Reply:
Thanks for the clarification. That looks very neat. Sorry for not understanding what you proposed at first. The dnn code can be a bit confusing for a newcomer. I will correct the implementation as per your suggestion.

Reply:
@dkurt I tried the solution you proposed and it didn't work. It seems like the readNetFromONNX runs fully without calling the setUp net function. And so the exception, as before gets thrown [here](https://github.com/opencv/opencv/blob/81fad9e2472af758130a15e0cd9f105bea9cc1f4/modules/dnn/src/onnx/onnx_importer.cpp#L860)

Reply:
If exception is thrown in other place (not during instantiation of layer), then it can't be fixed via "custom layers" approach.

For example, there is no way to fix INT64 tensors due missing INT64 support in OpenCV.